compiler/rustc_trait_selection/src/traits/select/confirmation.rs

```diff
@@ -955,6 +955,10 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                 // We already checked the compatibility of auto traits within `assemble_candidates_for_unsizing`.
                 let iter = data_a
                     .principal()
+                    .filter(|_| {
+                        // optionally drop the principal, if we're unsizing to no principal
+                        data_b.principal().is_some()
+                    })
```

Comment:
am a bit confused, is this the code responsible for `dyn Trait` to `dyn SuperTrait` coercions? Because if so, how does it work given that `iter` uses `data_a.principal()` :sweat_smile:

Reply:
No, this code is only responsible for auto trait dropping, lifetime changing, and now dropping principal. There's a different function for upcasting.

Reply:
Essentially this just drops a_principal if b_principal is gone.